src/commands/preview/proxy.rs

```diff
@@ -20,104 +18,175 @@ use url::Url;
 use crate::settings::global_user::GlobalUser;
 use crate::settings::target::Target;
 
-use crate::commands;
 use crate::commands::preview::upload;
 
 const PREVIEW_HOST: &str = "rawhttp.cloudflareworkers.com";
 
-pub fn proxy(
-    mut target: Target,
+#[derive(Clone)]
+struct ProxyConfig {
+    host: String,
+    listening_address: SocketAddr,
+    is_https: bool,
+}
+
+impl ProxyConfig {
+    pub fn new(
+        host: Option<&str>,
+        ip: Option<&str>,
+        port: Option<&str>,
+    ) -> Result<Self, failure::Error> {
+        let port: &str = match port {
+            Some(port) => port,
+            None => "8000",
+        };
+
+        let try_address = match ip {
+            Some(ip) => format!("{}:{}", ip, port),
+            None => format!("localhost:{}", port),
+        };
+
+        let mut address_iter = try_address.to_socket_addrs()?;
+
+        let listening_address = match address_iter.next() {
+            Some(ip) => Ok(ip),
+            None => Err(format_err!("Could not parse address {}", try_address)),
+        }?;
+
+        let host: String = match host {
+            Some(host) => host.to_string(),
+            None => "https://example.com".to_string(),
+        };
+
+        let parsed_url = match Url::parse(&host) {
+            Ok(host) => Ok(host),
+            Err(_) => Url::parse(&format!("https://{}", host)),
+        }?;
+
+        let scheme: &str = parsed_url.scheme();
+        if scheme != "http" && scheme != "https" {
+            failure::bail!("Your host scheme must be either http or https")
+        }
+        let is_https = scheme == "https";
+
+        let host = match parsed_url.host_str() {
+            Some(host_str) => Ok(host_str.to_string()),
+            None => Err(format_err!(
+                "Invalid host, accepted formats are http://example.com or example.com"
+            )),
+        }?;
+
+        let proxy = ProxyConfig {
+            listening_address,
+            host,
+            is_https,
+        };
+
+        Ok(proxy)
```

Comment:
I don't know what style wrangler uses, but in these cases, i tend to not introduce the `let` above, that is, I just write

    Ok(ProxyConfig {

but I can see things either way